Framework/Kernel/inc/MantidKernel/Cache.h

```diff
@@ -123,6 +167,7 @@ template <class KEYTYPE, class VALUETYPE> class DLLExport Cache {
     bool isValid = it_found != m_cacheMap.end();
 
     if (isValid) {
+      std::lock_guard<std::mutex> lock(m_mutex);
```

Comment:
My goal was to protect updates to the keys, but like you say above, this mutex alone is most likely not sufficient. 

Here's one [solution using atomics](http://stackoverflow.com/questions/15185856/tbb-concurrent-hash-maps-how-to-compare-and-set). Unfortunately, `BoundingBox` isn't trivially copyable, make this this solution more difficult to implement.